[jit] Add some of nn.init to weak script

driazati opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19640 [jit] Add some of nn.init to weak script**


Differential Revision: [D15065332](https://our.internmc.facebook.com/intern/diff/15065332/)

Commit:
[jit] Add some of nn.init to weak script

Commit:
Update on "[jit] Add some of nn.init to weak script"

[jit] Add some of nn.init to weak script

gh-metadata: pytorch pytorch 19640 gh/driazati/26/head

eellison commented on torch/nn/init.py:
```diff
@@ -105,10 +135,11 @@ def ones_(tensor):
-    with torch.no_grad():
-        return tensor.fill_(1)
+    # type: (Tensor) -> Tensor
+    # return _no_grad_fill_(tensor, 1.)
```
Accidental comment out

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -395,7 +395,48 @@ RegisterOperators reg(
+       push(stack, at::_th_normal_(tensor, mean, std));
+       return 0;
+     }),
+     Operator("aten::_no_grad_fill_(Tensor(a) tensor, float val) -> Tensor(a)", [](Stack& stack) {
```
It should be Tensor(a!) since it is mutated

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -395,7 +395,48 @@ RegisterOperators reg(
          "Tensor? unsorted_indices) -> (Tensor, Tensor, Tensor?, Tensor?)",
-         [](Stack& stack) { return 0; })
+         [](Stack& stack) { return 0; }),
+     Operator("aten::_no_grad_uniform_(Tensor(a) tensor, float a, float b) -> Tensor(a)", [](Stack& stack) {
```
This should be mutated as well right ?

eellison commented on torch/csrc/jit/register_special_ops.cpp:
```diff
@@ -395,7 +395,48 @@ RegisterOperators reg(
+       push(stack, at::fill_(tensor, val));
+       return 0;
+     }),
+     Operator("aten::_no_grad_zero_(Tensor(a) tensor) -> Tensor(a)", [](Stack& stack) {
```
mutated

eellison commented on torch/nn/init.py:
```diff
@@ -241,10 +276,11 @@ def xavier_normal_(tensor, gain=1):
-    std = gain * math.sqrt(2.0 / (fan_in + fan_out))
-    with torch.no_grad():
-        return tensor.normal_(0, std)
+    std = gain * math.sqrt(2.0 / float(fan_in + fan_out))
```
Why did you need the float cast ?

driazati replied:
`fan_in` and `fan_out` are ints, and there's no `aten::div(float, int)` currently

eellison replied:
Can you add that to this PR? it should be like a two line change - `DEFINE_INT_FLOAT_OP(aten::div, a / b, float)` in register_prim_ops and add '/' to test_number_math.

eellison replied:
This wasn't really resolved.

driazati replied:
At least should be left for a follow up, I feel like it counts as an implicit `int` -> `float` conversion which if we're going to add should be at the compiler level instead of per-op.

eellison commented on test/test_jit.py:
```diff
@@ -11193,6 +11193,34 @@ def nest_for_ret(x):
+            # Compare functions
+            init_fn = getattr(torch.nn.init, name)
+            script_out = self.runAndSaveRNG(cu.test, args_fn())
+            eager_out = self.runAndSaveRNG(init_fn, args_fn())
```
Can you add an assertion that there is no python op

eellison commented on test/test_jit.py:
```diff
@@ -11193,6 +11193,34 @@ def nest_for_ret(x):
+            arg_str = ', '.join([chr(i + ord('a')) for i in range(len(args_fn()))])
+
+            code = dedent('''
+                def test({}):
```
nit: a little more readable if you name the formatting slots

Commit:
Update on "[jit] Add some of nn.init to weak script"

[jit] Add some of nn.init to weak script

gh-metadata: pytorch pytorch 19640 gh/driazati/26/head

eellison commented on test/test_jit.py:
```diff
@@ -11193,6 +11193,36 @@ def nest_for_ret(x):
 
+    def test_nn_init(self):
+        tests = (
+            ('constant_', (lambda: (torch.ones(2, 2), 2.5)), "Tensor, float"),
```
Can you switch it to rand ?

eellison reviewed (approved):
LGTM. A little more coverage on different inputs would be nice, as would be adding the div float / int, which should be a really really easy change.

eellison commented on torch/nn/init.py:
```diff
@@ -105,11 +136,12 @@ def ones_(tensor):
     """
-    with torch.no_grad():
-        return tensor.fill_(1)
+    return _no_grad_fill_(tensor, 1.)
```
Hmm it looks like this is changing the semantics, because previously the type of the tensor would have been preserved, but now it is being filled with floats
